Online VM Snapshot with guest agent

ShellyKa13 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
This PR integrates guest agent with online vm snapshot.
In case guest agent is available and the snapshot is taken when the VM is running the guest fs will be frozen before the snapshot
is taken an unfrozen after.
The freeze/unfreeze api is exposed via virt handler endpoint which tunnels the command to the virt launcher which sends it using the guest agent to the guest.
*working on adding some functional tests

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Integrate guest agent to online VM snapshot

```

mhenriks reviewed (commented):
So far so good!  Great start!

ShellyKa13 commented on pkg/virt-controller/watch/snapshot/snapshot.go:
```diff
@@ -559,6 +583,13 @@ func (ctrl *VMSnapshotController) getVolumeSnapshotClass(storageClassName string
 	return "", fmt.Errorf("%d matching VolumeSnapshotClasses for %s", len(matches), storageClassName)
 }
 
+func (ctrl *VMSnapshotController) updateVMSnapshotError(vmSnapshot *snapshotv1.VirtualMachineSnapshot, reason string) {
```
I havn't returned here and error cause I figured it is being called when there was already an error, so if we fail to updated the error on the vmsnapshot it just another error which we have nothing to do but to update on the vmsnapshot and return it, so we already return another probably more important error

mhenriks replied:
The caller can always choose to ignore the error returned.  Or just log it.

ShellyKa13 replied:
ok sounds good

ShellyKa13 commented on pkg/virt-controller/watch/snapshot/snapshot.go:
```diff
@@ -821,13 +926,32 @@ func checkVMRunning(vm *kubevirtv1.VirtualMachine) (bool, error) {
 
-func (ctrl *VMSnapshotController) checkVMIRunning(vm *kubevirtv1.VirtualMachine) (bool, error) {
+func (ctrl *VMSnapshotController) getVMSnapshot(vmSnapshotContent *snapshotv1.VirtualMachineSnapshotContent) (*snapshotv1.VirtualMachineSnapshot, error) {
+	vmSnapshotName := *vmSnapshotContent.Spec.VirtualMachineSnapshotName
```
ok will add a check

ShellyKa13 commented on pkg/virt-controller/watch/snapshot/snapshot.go:
```diff
@@ -218,6 +219,18 @@ func (ctrl *VMSnapshotController) updateVMSnapshotContent(content *snapshotv1.Vi
+				if err != nil {
+					return 0, err
+				}
+				err = ctrl.freezeGuestFSIfNeeded(vmSnapshot)
```
I am not sure, supposedly the call is immediate, sends the command which should happen and return, I'm not sure if with busy fs it might take more time. I don't think we should call the freeze, continue with other snapshots and return to this snapshot when it completes, but i'm not sure.. how can we check? what do you think is best?

mhenriks replied:
I guess let's not get too complicated now.  I think it would be wise to increase snapshot controller threads here:  https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/application.go#L430.  Maybe double?

Should also think about a way to log/track how long that call is taking.

ShellyKa13 replied:
ok sure

ShellyKa13 replied:
@mhenriks I fixed all the other comments and add two commits one for functional tests and another for the increase of the number of the threads and added logging of time for the freeze/unfreeze, from what I saw the freeze unfreeze of vms in the tests took around 100ms to freeze and 25ms to unfreeze

ShellyKa13 commented:
/retest

ShellyKa13 commented:
/retest

mhenriks commented on pkg/virt-controller/watch/snapshot/snapshot.go:
```diff
@@ -645,6 +677,85 @@ func (ctrl *VMSnapshotController) updateSnapshotStatus(vmSnapshot *snapshotv1.Vi
+	vmName := vmSnapshot.Spec.Source.Name
+	log.Log.V(3).Infof("Freezing vm %s file system before taking the snapshot", vmName)
+
+	defer timeTrack(time.Now(), "Freezing vm")
```
this time span will also include `updateVMSnapshotError` or `replaceGuestAgentIndication`

ShellyKa13 replied:
I know.. didnt think its a problem, but ill change it to only wrap the freeze call

ShellyKa13 commented:
/retest

mhenriks commented on pkg/virt-controller/watch/application.go:
```diff
@@ -81,7 +81,7 @@ const (
 	ephemeralDiskDir = virtShareDir + "-ephemeral-disks"
 
-	defaultControllerThreads    = 3
+	defaultControllerThreads    = 6
```
Let's only increase threads for the snapshot controller.

mhenriks commented on tests/storage/restore.go:
```diff
@@ -742,7 +742,60 @@ var _ = SIGDescribe("[Serial]VirtualMachineRestore Tests", func() {
+
+			})
+
+			It("should restore a vm from an online snapshot with guest agent", func() {
```
How about another case where we snapshot/restore boot disk?

rmohr reviewed (commented):
Great to see the progress on the PR. I only looked at the PR superficial, but it looks to me that the VMI does not indicate when the guest agent freeze command was invoked. Can it be that this is missing?

If so, can there be cases where e.g. the snapshot controller has fundamental issues, and VMs are therefore not unfrozen, but we don't see it in the VMI? With other words, would the VMI look perfectly healthy from the outside? Also, would it make sense to indicate on the VMI in general somehow that currently a snapshot is invoked which may have impact on the functionality?

ShellyKa13 commented:
@rmohr In a previous PR that was already merged there is a separate mechanism in the virt-handler and launcher which polls for the guest fsfreeze status, so in case the fs is frozen it will mark it on the vmi object and once unfrozen it will remove the frozen label.
Also there is a mark that snapshot is in progress.
Furthermore one of my next tasks is to create a mechanism that incase there is something wrong with the snapshot controller and the vm is froazen for more the X time then the vm will be unfrozen by the virt handler.

rmohr commented:
> @rmohr In a previous PR that was already merged there is a separate mechanism in the virt-handler and launcher which polls for the guest fsfreeze status, so in case the fs is frozen it will mark it on the vmi object and once unfrozen it will remove the frozen label.
> Also there is a mark that snapshot is in progress.
> Furthermore one of my next tasks is to create a mechanism that incase there is something wrong with the snapshot controller and the vm is froazen for more the X time then the vm will be unfrozen by the virt handler.

I see thanks. You probably talk about https://github.com/kubevirt/kubevirt/pull/5777. Can you please add dedicated e2e tests just for freeze, since there were no tests added in 5777? @mhenriks FYI.

ShellyKa13 commented:
@rmohr in the PR #5777 I only added the update of fsfreezestatus to the vmi object without any actual action of freeze, thats why I only added UT. In this PR I added the integration of the guest agent to the vm snapshot. In this PR I added tests for online snapshot with guest agent which freeze and unfreeze the vm during the snapshot. I can write in a different PR a test that runs freeze and unfreeze inside the virt-launcher pod using the guest agent and only check that the vmi object fsfreeze status was updated but I thought the UT in the previous PR basically covered all the parts in this test.

mhenriks commented:
I feel really great about the state of this PR.  Great work @ShellyKa13!  Regarding @rmohr's requested functional test.  I am alright if it goes into another PR.

mhenriks commented:
@ShellyKa13 after thinking about it, maybe it would be possible to catch frozen VMI status by temporarily putting a dummy webhook on VolumeSnapshots?

ShellyKa13 commented:
@mhenriks I moved the unfreeze of the vm from the update snapshot status to the main reconcile loop and added a test which checks that the vm gets unfrozen even if deleted before completion, this test also check the vm fsfreezestatus as requested by @rmohr. check new last 2 commits

mhenriks commented:
Latest commits look good.  Not sure what's up with the test failures though.

mhenriks commented on tests/storage/snapshot.go:
```diff
@@ -154,6 +217,316 @@ var _ = SIGDescribe("[Serial]VirtualMachineSnapshot Tests", func() {
+						updatedVMI.Status.FSFreezeStatus == "frozen"
+				}, time.Minute, 2*time.Second).Should(BeTrue())
+
+				deleteSnapshot()
```
maybe wait for snapshot to actually be gone (get returns not found) here?

mhenriks commented:
/retest

ShellyKa13 commented:
/retest

ShellyKa13 commented:
/retest

ShellyKa13 commented:
/retest

mhenriks commented:
/lgtm
/approve
